src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -157,25 +157,24 @@ inline int dd::SingleThreadSampler::draw_sample(Variable &variable,
       proposal = -1;
 
       // calculate potential for each proposal
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        int propose = variable.get_domain(i);
         if (is_free_sample) {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<true>(variable, propose);
         } else {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<false>(variable, propose);
         }
-        sum = logadd(sum, varlen_potential_buffer[propose]);
+        sum = logadd(sum, varlen_potential_buffer[i]);
       }
 
       // flip a coin
       *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
-        acc += exp(varlen_potential_buffer[propose] - sum);
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        acc += exp(varlen_potential_buffer[i] - sum);
         if (*this->p_rand_obj_buf <= acc) {
-          proposal = propose;
+          proposal = variable.get_domain(i);
```

Comment:
I think this is the only place where the domain is ever iterated, apart from where it's printed, and the order isn't important I think while printing informational messages.  If performance/efficiency matter I think repeating code is worth doing.  If you think such abstraction is absolutely important, we can surely create one that accepts a lambda.  But that abstraction would incur overhead, and I'm not sure we want that.  The multinomial vs. boolean specialization could also be viewed as code repetition, but I think we're fine with that, right?  I'd argue for repeating some amount of code for dense/sparse within multinomial for the same reason.